src/Service/TrustedDevice/TrustedDeviceService.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+declare(strict_types = 1);
+
+/**
+ * Copyright 2024 SURFnet bv
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\Tiqr\Service\TrustedDevice;
+
+use Exception;
+use Psr\Log\LoggerInterface;
+use Surfnet\Tiqr\Service\TrustedDevice\DateTime\ExpirationHelperInterface;
+use Surfnet\Tiqr\Service\TrustedDevice\Exception\CookieNotFoundException;
+use Surfnet\Tiqr\Service\TrustedDevice\Exception\DecryptionFailedException;
+use Surfnet\Tiqr\Service\TrustedDevice\Exception\InvalidAuthenticationTimeException;
+use Surfnet\Tiqr\Service\TrustedDevice\Http\CookieHelperInterface;
+use Surfnet\Tiqr\Service\TrustedDevice\ValueObject\CookieValue;
+use Surfnet\Tiqr\Service\TrustedDevice\ValueObject\CookieValueInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects) | Coupling is high as we are integrating logic into the infrastructure
+ */
+class TrustedDeviceService
+{
+    public function __construct(
+        private readonly CookieHelperInterface $cookieHelper,
+        private readonly ExpirationHelperInterface $expirationHelper,
+        private readonly LoggerInterface $logger,
+    ) {
+    }
+
+    public function registerTrustedDevice(Response $response, string $userId, string $notificationAddress): void
+    {
+        $this->store($response, CookieValue::from($userId, $notificationAddress));
+    }
+
+    public function isTrustedDevice(
+        CookieValueInterface $cookie,
+        string $userId,
+        string $notificationAddress,
+    ): bool {
+        if (!$this->isCookieValid($cookie, $userId, $notificationAddress)) {
+            return false;
+        }
+
+        $this->logger->notice('Push-notification MFA is allowed for the device based on the cookie.');
+        return true;
+    }
+
+
+    private function store(Response $response, CookieValueInterface $cookieValue): void
+    {
+        $this->cookieHelper->write($response, $cookieValue);
+    }
+
+    public function read(Request $request, string $userId, string $notificationAddress): ?CookieValueInterface
```

Comment:
1. Have you considered to make the return type null-safe? So basically create a NullCookieValue object instead of returning null?
2. Nitpick: move public methodes up in the class

Reply:
1. Discussed. Yes, but no.
2. :+1: